src/coreclr/jit/fgprofile.cpp

```diff
@@ -887,130 +1628,821 @@ PhaseStatus Compiler::fgIncorporateProfileData()
+    {
+        BasicBlock::weight_t profileWeight;
+
+        if (fgGetProfileWeightForBasicBlock(block->bbCodeOffs, &profileWeight))
+        {
+            fgSetProfileWeight(block, profileWeight);
+        }
+    }
+}
+
+//------------------------------------------------------------------------
+// EfficientEdgeCountReconstructor: reconstruct block counts from sparse
+//   edge counts.
+//
+// Notes:
+//    The algorithm is conceptually simple, but requires a bit of bookkeeping.
+//
+//    First, we should have a correspondence between the edge count schema
+//    entries and the non-tree edges of the spanning tree.
+//
+//    The instrumentation schema may be partial, if any importer folding was
+//    done. Say for instance we have a method that is ISA sensitive to x64 and
+//    arm64, and we instrument on x64 and are now jitting on arm64. If so
+//    there may be missing schema entries. If we are confident the IL and
+//    jit IL to block computations are the same, these missing entries can
+//    safely be presumed to be zero.
+//
+//    Second, we need to be able to reason about the sets of known and
+//    unknown edges that are incoming and outgoing from any block. These
+//    may not quite be the edges we'd see from iterating successors or
+//    building pred lists, because we create special pseudo-edges during
+//    instrumentation. So, we also need to build up data structures
+//    keeping track of those.
+//
+//    Solving is done in four steps:
+//    * Prepare
+//      *  walk the blocks settting up per block info, and a map
+//         for block schema keys to blocks.
+//      * walk the schema to create info for the known edges, and
+//         a map from edge schema keys to edges.
+//    * Evolve Spanning Tree
+//      * for non-tree edges, presume any missing edge is zero
+//        (and hence, can be ignored during the solving process
+//      * for tree edges, verify there is no schema entry, and
+//        add in an unknown count edge.
+//    * Solve
+//      * repeatedly walk blocks, looking for blocks where all
+//        incoming our outgoing edges are known. This determines
+//        the block counts.
+//      * for blocks with known counts, look for cases where just
+//        one incoming or outgoing edge is unknown, and solve for
+//        them.
+//    * Propagate
+//      * update block counts. bail if there were errors.
+//        * mark rare blocks, and special case handler entries
+//        * (eventually) try "fixing" counts
+//      * (eventually) set edge likelihoods
+//      * (eventually) normalize
+//
+//   If we've done everything right, the solving is guaranteed to
+//   converge.
+//
+//   Along the way we may find edges with negative counts; this
+//   is an indication that the count data is not self-consistent.
+//
+class EfficientEdgeCountReconstructor : public SpanningTreeVisitor
+{
+private:
+    Compiler*     m_comp;
+    CompAllocator m_allocator;
+    uint32_t      m_blocks;
```

Comment:
why are you using uint32_t / int32_t here? Can't we use the normal C++ types?

Reply:
Sure. There will be a bit left as some of it comes from the instrumentation schema declaration in corjit.h